src/peer_select.cc

```diff
@@ -582,48 +717,127 @@ PeerSelector::selectSomeNeighbor()
     }
 
 #if USE_CACHE_DIGESTS
-    if ((p = neighborsDigestSelect(request))) {
-        if (neighborType(p, request->url) == PEER_PARENT)
-            code = CD_PARENT_HIT;
-        else
-            code = CD_SIBLING_HIT;
-    } else
+    neighborsDigestSelect(this, request);
 #endif
-        if ((p = netdbClosestParent(request))) {
-            code = CLOSEST_PARENT;
-        } else if (peerSelectIcpPing(request, direct, entry)) {
-            debugs(44, 3, "Doing ICP pings");
-            ping.start = current_time;
-            ping.n_sent = neighborsUdpPing(request,
-                                           entry,
-                                           HandlePingReply,
-                                           this,
-                                           &ping.n_replies_expected,
-                                           &ping.timeout);
-
-            if (ping.n_sent == 0)
-                debugs(44, DBG_CRITICAL, "WARNING: neighborsUdpPing returned 0");
-            debugs(44, 3, ping.n_replies_expected <<
-                   " ICP replies expected, RTT " << ping.timeout <<
-                   " msec");
-
-            if (ping.n_replies_expected > 0) {
-                entry->ping_status = PING_WAITING;
-                eventAdd("PeerSelector::HandlePingTimeout",
-                         HandlePingTimeout,
-                         this,
-                         0.001 * ping.timeout,
-                         0);
-                return;
-            }
+    netdbClosestParent(this, request);
+}
+
+static void
+checkNextPingNeighborAccessWrapper(allow_t answer, void *data)
+{
+    PeerSelector *selector = static_cast<PeerSelector *>(data);
+    selector->checkNextPingNeighborAccess(answer);
+}
+
+#if 0
```

Comment:
The PeerSelector::checkNextPingNeighborAccess calls the selectServersToPing and selectServersToPing again the checkNextPingNeighborAccess while checking acls for Peers. If only nonblocking ACLs are used then may result to many recursive calls depending the number the peers.
I do not know how big problem can be so maybe it is good idea to split this procedure, using AsyncCalls.

Reply:
> If only nonblocking [Alex: I assume you meant "fast" here] ACLs are used then may result to many recursive calls depending the number the peers.

I do not think this is about fast ACLs -- nonBlockingCheck() is always asynchronous, regardless of whether the ACLs are fast or slow, right? This is about cached checks -- aclPeersCache hits are (currently) synchronous. I suggest adding a comment to the doIfAllowed() used in my sketch: `TODO: Decide whether to make cache hits asynchronous, reducing recursion and improving exception handling`

Reply:
> I do not think this is about fast ACLs -- nonBlockingCheck() is always asynchronous, regardless of whether the ACLs are fast or slow, right?

The nonBlockingCheck may return to the same call, if all acls ar fast, or they have cached results. If you have lets say 10 servers to ping you may result to 10*AclCheckRequiredRecursiveCalls recursive function calls.

Reply:
> The nonBlockingCheck may return to the same call

I stand corrected. Moreover, nonBlockingCheck() probably does not use AsyncCalls so the part of my comment about exception safety improvements is also wrong. Sorry.

Where the code needs to find the first allowed peer, we may be able to add a hack that allows nonBlockingCheck() callback to detect synchronous nonBlockingCheck() execution and avoid recursion -- the callback will know that the nonBlockingCheck()-calling loop "above" it will iterate again, so there is no need to recurse. However, let's simply recurse for now so that we can implement this tricky "optimization" in isolation from the core changes.

```
// TODO: Avoid deep recursion when finding the first allowed peer using fast/cached ACLs.
... nonBlockingCheck(...)
```